lib/mongo/retryable.rb

```diff
@@ -15,463 +15,38 @@
-          raise original_error
-        end
-        e.add_note("attempt 2")
-        raise e
-      rescue Error, Error::AuthError => e
-        e.add_note('modern retry')
-        original_error.add_note("later retry failed: #{e.class}: #{e}")
-        raise original_error
-      end
-    rescue Error::RaiseOriginalError
-      raise original_error
-    end
-
-    def retry_write(original_error, session, txn_num, &block)
-      # We do not request a scan of the cluster here, because error handling
-      # for the error which triggered the retry should have updated the
-      # server description and/or topology as necessary (specifically,
-      # a socket error or a not master error should have marked the respective
-      # server unknown). Here we just need to wait for server selection.
-      server = select_server(cluster, ServerSelector.primary, session)
-      unless server.retry_writes?
-        # Do not need to add "modern retry" here, it should already be on
-        # the first exception.
-        original_error.add_note('did not retry because server selected for retry does not supoprt retryable writes')
-
-        # When we want to raise the original error, we must not run the
-        # rescue blocks below that add diagnostics because the diagnostics
-        # added would either be rendundant (e.g. modern retry note) or wrong
-        # (e.g. "attempt 2", we are raising the exception produced in the
-        # first attempt and haven't attempted the second time). Use the
-        # special marker class to bypass the ordinarily applicable rescues.
-        raise Error::RaiseOriginalError
-      end
-      log_retry(original_error, message: 'Write retry')
-      yield(server, txn_num, true)
-    rescue Error::SocketError, Error::SocketTimeoutError => e
-      e.add_note('modern retry')
-      e.add_note('attempt 2')
-      raise e
-    rescue Error::OperationFailure => e
-      e.add_note('modern retry')
-      if e.label?('RetryableWriteError')
-        e.add_note('attempt 2')
-        raise e
-      else
-        original_error.add_note("later retry failed: #{e.class}: #{e}")
-        raise original_error
-      end
-    rescue Error, Error::AuthError => e
-      # Do not need to add "modern retry" here, it should already be on
-      # the first exception.
-      original_error.add_note("later retry failed: #{e.class}: #{e}")
-      raise original_error
-    rescue Error::RaiseOriginalError
-      raise original_error
-    end
-
-    # This is a separate method to make it possible for the test suite to
-    # assert that server selection is performed during retry attempts.
-    def select_server(cluster, server_selector, session)
+    # @return [ Mongo::Server ] A server matching the server preference.
+    def select_server(cluster, server_selector, session, failed_server = nil)
       retried_server_selector = false
       begin
         server_selector.select_server(cluster, nil, session)
```

Comment:
This removes an upstream change passing `deprioritized`

![image](https://github.com/braze-inc/mongo-ruby-driver/assets/6033/7ab9488f-1827-4265-b8b1-2f786cd1f29e)

I would think we want to keep that,
and maybe set `failed_server = nil` after the `cluster.scan!`